File: ExoPi/ExoPi/Valve.py
import gpiozero as gp
import time
class Valve(object):
    """description of class"""
    def __init__(self,name,index,valveRecQue,valveRecLock):
        self.name = name
        self.pin = gp.OutputDevice(index)
        self.valveRecQue = valveRecQue
        self.valveRecLock = valveRecLock
        self.onStr = ','+name + ',1'
        self.offStr = ','+name + ',0'
    def on(self):
        self.pin.on()
        curTime = time.time()
        with self.valveRecLock:
            self.valveRecQue.put(str(curTime)+self.onStr)
    def off(self):
        self.pin.off()
        curTime = time.time()
        with self.valveRecLock:
            self.valveRecQue.put(str(curTime)+self.offStr)

File: Exo/ExoMain/ExoMain/Script/src/Valve.cpp
#include "Valve.h"
#include <array>
using namespace std;
void Valve::On(int curTime){
    
    this->gpioPin->On();

    array<bool,1> curRes={true};
    this->valveRec->PushData((unsigned long)curTime,curRes);
   
}
void Valve::Off(int curTime){
    this->gpioPin->Off();
    //this->pin->Off();
    array<bool,1> curRes = {false};
    this->valveRec->PushData((unsigned long)curTime,curRes);

}

Valve::Valve(string name,string path, int valveId,int _valIdx)
{
    
    this->valIdx = _valIdx;
    this->name = name;
    
    this->gpioPin.reset(new Pin(valveId));
    
    this->valveRec=new Recorder<bool,1>(this->name,path,"time,"+this->name);
    
}

Valve::~Valve()
{
    this->gpioPin->Off();

    if(this->dummy)
        delete this->valveRec;
    
}
string Valve::GetValveName(){
    return this->name;
}
int Valve::GetValIdx(){
    return this->valIdx;
}